### PR TITLE
Improve CodeEdit to allow folding comment blocks

### LINKED ISSUE #8538: Add support for foldable comment blocks
### Describe the project you are working on

folding_comments
 Code editor should have foldable comments to allow working with lengthy comments

### Describe the problem or limitation you are having in your project

none

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Enhance user experience with code editor in godot

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

this pull request explores implementation 
 - [SeremTitus:folding-doc-comments](https://github.com/godotengine/godot/pull/85453)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

no, the API to achieve this feature are not exposed

### Is there a reason why this should be core and not an add-on in the asset library?

 the API to achieve this feature are not exposed

### PR DESCRIPTION
**Pull request by SeremTitus**
folding_comments
allows foldable comments and extensible by taking prefix of a desired code block to make it foldable

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/8538*

### CODE REVIEW

**Comment by Chaosus** [comment]
You need to squash the commits to a single one (see https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request)

File: scene/gui/code_edit.cpp [hunk 1]
```diff
@@ -742,6 +742,48 @@ void CodeEdit::_backspace_internal(int p_caret) {
...
+
+void CodeEdit::set_fold_blocks_prefixes(const PackedStringArray &p_prefixes) {
+	fold_blocks_prefixes.clear();
+	for (int i = 0; i < p_prefixes.size(); i++) {
+		const String prefix = p_prefixes[i];
+		if (fold_blocks_prefixes.has(prefix) || prefix == "") {
+			continue;
+		}
+		fold_blocks_prefixes.append(prefix);
+	}
```
Comment (AThousandShips) [inline scene/gui/code_edit.cpp:780]:
```suggestion
	for (const String &prefix : p_prefixes) {
		if (prefix.is_empty() || fold_blocks_prefixes.has(prefix)) {
			continue;
		}
		fold_blocks_prefixes.append(prefix);
	}
```
Use a foreach loop, and put the simpler check first

File: scene/gui/code_edit.cpp [hunk 2]
```diff
@@ -742,6 +742,48 @@ void CodeEdit::_backspace_internal(int p_caret) {
...
+			continue;
+		}
+		// this line and next  but not previous begins with fold block prefix
+		if (get_line(p_line).begins_with(fold_blocks_prefixes[i]) && get_line(p_line + 1).begins_with(fold_blocks_prefixes[i])) {
+			if (get_line(p_line - 1).begins_with(fold_blocks_prefixes[i])) {
+				return false;
+			}
+			return true;
+		}
+	}
```
Comment (AThousandShips) [inline scene/gui/code_edit.cpp:760]:
```suggestion
	for (const String &prefix : fold_blocks_prefixes) {
		// Empty string prefix will always return true when begins_with is checked.
		if (prefix.is_empty()) {
			continue;
		}
		// This line and next, but not previous, begins with fold block prefix.
		if (get_line(p_line).begins_with(prefix) && get_line(p_line + 1).begins_with(prefix)) {
			return !get_line(p_line - 1).begins_with(prefix);
		}
	}
```
Some simplifications and style details

File: scene/gui/code_edit.cpp [hunk 3]
```diff
@@ -1620,6 +1667,21 @@ void CodeEdit::fold_line(int p_line) {
...
 
+	// fold Folding Block
+	if (!is_line_code_region_start(p_line) && _is_fold_blocks_start(p_line)) {
+		String fold_blocks_prefix;
+		for (int i = 0; i < fold_blocks_prefixes.size(); i++) {
+			if (get_line(p_line).begins_with(fold_blocks_prefixes[i])) {
+				fold_blocks_prefix = fold_blocks_prefixes[i];
+				break;
+			}
+		}
```
Comment (AThousandShips) [inline scene/gui/code_edit.cpp:1678]:
```suggestion
		for (const String &prefix : fold_blocks_prefixes) {
			if (get_line(p_line).begins_with(prefix)) {
				fold_blocks_prefix = prefix;
				break;
			}
		}
```

**Comment by SeremTitus** [comment]
> You need to squash the commits to a single one (see https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request)

 squashed

**Comment by Calinou** [comment]
Thanks for opening a pull request :slightly_smiling_face: 

Feature pull requests should be associated to a *feature proposal* to justify the need for implementing the feature in Godot core. Please open a proposal on the [Godot proposals](https://github.com/godotengine/godot-proposals) repository (and link to this pull request in the proposal's body).

This feature looks interesting and I've seen it in a few other editors, although Visual Studio Code doesn't seem to feature this (or it's not enabled by default).

**Comment by AThousandShips** [comment]
I'd suggest building locally to find the errors in your code instead of taxing the CI with it, at least until you've ironed out some things

**Comment by AThousandShips** [comment]
Are you unable to build locally? You will be much more able to find errors in your code that way (and you aren't still having several errors in your code that were picked up by CI that remain) 🙂

**Comment by SeremTitus** [comment]
@AThousandShips , am adding minimum size which has come with sum errors am sure I can iron the out in a couple of days...
And yes I can and do build locally it quite slow I can to seem to figure how to create a cache using scons,I prefer the CI but I will keep minimal use to this branch

**Comment by Yagich** [comment]
Is this not already a feature? On 4.2rc2, I have this:
image

**Comment by SeremTitus** [comment]
@Yagich it might be Easter egg

**Comment by Yagich** [comment]
> it might be an Easter egg

How so? It works as expected, and in line with the proposal you made.

**Comment by SeremTitus** [comment]
> > it might be an Easter egg
> 
> How so? It works as expected, and in line with the proposal you made.

It is a good mystery, I don't really know

**Comment by SeremTitus** [comment]
Does this feature exist in master?

**Comment by AThousandShips** [comment]
Go ahead and check in the release 🙂, see [here](https://github.com/godotengine/godot/releases/tag/4.2-stable)

**Comment by SeremTitus** [comment]
It  does it was introduce some where before 4.2.beta6, I do not know how I missed it . Am not sure when where how it was implimented though

**Comment by AThousandShips** [comment]
Already implemented by:
* https://github.com/godotengine/godot/pull/74843

I would say

This can be closed as the functionality is already present, if some improvements or other details can be added from this please open a new PR with those features 🙂 

Thank you for your contribution nonetheless

**Comment by SeremTitus** [comment]
achieved by add_comment_delimiter("#","",)

**Pull request closed without merging**